### PR TITLE
add wait_tensor to ops to skip in DCE pass

### PR DESCRIPTION
**Pull request by bdhirsh**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126784
* __->__ #126541
* #127325

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Can we put accurate schema on this op?

**Comment by bdhirsh** [comment]
Filed an issue for it here: https://github.com/pytorch/pytorch/issues/126773 (it's a bit annoying because we'd need to either tell functionalization to explicitly not functionalize the op, or add a functional variant + defunctionalize it in inductor)

File: torch/distributed/_tensor/api.py [hunk 1]
```diff
@@ -773,3 +773,8 @@ def replicate_module_params_buffers(m: nn.Module, mesh: DeviceMesh) -> None:
             )
 
     return module
+
+
+# wait_tensor is only available when we initialize distributed.
+# Make sure it cannot be DCE'd.
+torch.fx.node._side_effectful_functions.add(torch.ops._c10d_functional.wait_tensor)
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:780]: can this be put into the `_functional_collectives.py` instead?

**Comment by fegin** [comment]
This PR is no longer required as functional collective will add wait_tensor to the _side_effectful_functions during the module initialization.

**Pull request closed without merging**